cargo test then cargo tarpaulin causes BorrowMut error and linking failure

So when running cargo test with default rust flags and then running cargo tarpaulin (version 0.18.0-alpha1) cargo can output the following error (it's called within tarpaulin)

```
error: Broken pipe (os error 32)
warning: build failed, waiting for other jobs to finish...
thread 'main' panicked at 'already borrowed: BorrowMutError', src/tools/cargo/src/cargo/util/config/mod.rs:307:20
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
Mar 01 19:28:55.644 ERROR cargo_tarpaulin: Failed to compile tests! Error: signals: linking with `cc` failed: exit code: 1
Error: "Failed to compile tests! Error: signals: linking with `cc` failed: exit code: 1"
```

To recreate the error given an example project (https://github.com/Smithay/calloop) we do the following:

```
cargo install cargo-tarpaulin --version 0.18.0-alpha1
git clone --depth 1 https://github.com/Smithay/calloop 
cd calloop
cargo test
cargo tarpaulin
```

One salient note cargo tarpaulin changes the rust flags to `RUSTFLAGS=" -C link-dead-code -C debuginfo=2 --cfg=tarpaulin "` so I wonder if this is a case of incremental compilation going wrong?

Further comment from the person who raised this with me in tarpaulin initially:

> if we do the reverse (cargo clean, then cargo tarpaulin, then cargo test), then cargo test fails with a more detailed error. For example on calloop I get undefined reference to core::ptr::drop_in_place'`.

--- a/src/cargo/core/compiler/job_queue.rs
+++ b/src/cargo/core/compiler/job_queue.rs
@@ -748,13 +748,14 @@ impl<'cfg> DrainState<'cfg> {
             }
         }
         if cx.bcx.build_config.emit_json() {
+            let mut shell = cx.bcx.config.shell();
             let msg = machine_message::BuildFinished {
                 success: error.is_none(),
             }
             .to_json_string();
-            if let Err(e) = writeln!(cx.bcx.config.shell().out(), "{}", msg) {
+            if let Err(e) = writeln!(shell.out(), "{}", msg) {
                 if error.is_some() {
-                    crate::display_error(&e.into(), &mut cx.bcx.config.shell());
+                    crate::display_error(&e.into(), &mut shell);
                 } else {
                     return Some(e.into());
                 }

